src/Ocelot.Provider.Consul/PollConsul.cs

```diff
@@ -1,53 +1,73 @@
 ﻿using Ocelot.Logging;
 using Ocelot.ServiceDiscovery.Providers;
-using Ocelot.Values;
 using System;
 using System.Collections.Generic;
+using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using Ocelot.Values;
 
-namespace Ocelot.Provider.Consul;
-
-public sealed class PollConsul : IServiceDiscoveryProvider, IDisposable
+namespace Ocelot.Provider.Consul
 {
-    private readonly IOcelotLogger _logger;
-    private readonly IServiceDiscoveryProvider _consulServiceDiscoveryProvider;
-    private Timer _timer;
-    private bool _polling;
-    private List<Service> _services;
-
-    public PollConsul(int pollingInterval, IOcelotLoggerFactory factory, IServiceDiscoveryProvider consulServiceDiscoveryProvider)
+    public sealed class PollConsul : IServiceDiscoveryProvider, IDisposable
     {
-        _logger = factory.CreateLogger<PollConsul>();
-        _consulServiceDiscoveryProvider = consulServiceDiscoveryProvider;
-        _services = new List<Service>();
+        private readonly IOcelotLogger _logger;
+        private readonly IServiceDiscoveryProvider _consulServiceDiscoveryProvider;
+
+        private readonly int _pollingInterval;
+        private DateTime _lastUpdateTime;
+
+        private List<Service> _services;
+        private readonly SemaphoreSlim _semaphore = new(1, 1);
 
-        _timer = new Timer(async x =>
+        public PollConsul(int pollingInterval, string serviceName, IOcelotLoggerFactory factory, IServiceDiscoveryProvider consulServiceDiscoveryProvider)
         {
-            if (_polling)
+            _logger = factory.CreateLogger<PollConsul>();
+            _consulServiceDiscoveryProvider = consulServiceDiscoveryProvider;
+            _pollingInterval = pollingInterval;
+            ServiceName = serviceName;
+            _services = new List<Service>();
+
+            //initializing with update time = DateTime.MinValue
+            //polling will occur then.
+            _lastUpdateTime = DateTime.MinValue;
+        }
+
+        public string ServiceName { get; }
+
+        /// <summary>
+        /// Getting the services, but, if first call,
+        /// retrieving the services and then starting the timer.
+        /// </summary>
+        /// <returns>the service list.</returns>
+        public async Task<List<Service>> Get()
+        {
+            await _semaphore.WaitAsync();
```

Comment:
Doesn't simple **lock**-section work here?
From the instance being initialized by `new(1, 1)` I see that you limit access for 1 concurrent thread.
Why not to use just **lock**-operator?

After that, I think, you can remove inheritance of the `IDisposable` interface.

Reply:
@raman-m I don't think you can use a lock in a Task

Reply:
Ok, so, I removed IDisposable and used lock instead. I also fixed the tests, they are passing now.